packages/authx/src/util/ratelimiter.ts

```diff
@@ -0,0 +1,62 @@
+import { performance } from "perf_hooks";
+import { TooManyRequests } from "../errors";
+
+/**
+ * Applies a simple in-memory rate limiting scheme. This system is designed to prevent a single
+ * malfunctioning client from bringing down the service. This system is only designed to prevent
+ * unintentional abuse by malfunctioning clients.
+ */
+export interface RateLimiter {
+  /**
+   * Applies the configured rate limit to the given key. If the key is being excessively used, throws a TooManyRequests exception.
+   * @param key A string representing the key to use for rate limiting. This is generally the id of an authorization,
+   *            client, or credential.
+   */
+  limit(key: string): void;
+}
+
+export class LocalMemoryRateLimiter implements RateLimiter {
+  private readonly map: { [key: string]: number[] } = {};
+
+  constructor(
+    private readonly limitPerWindow = 100,
+    private readonly window = 60 * 1_000,
+    private readonly timeSource: () => number = performance.now
+  ) {}
+
+  limit(key: string): void {
+    const currentTime = this.timeSource();
+
+    for (const existingKey of Object.keys(this.map)) {
+      for (let i = 0; i < this.map[existingKey].length; ++i) {
+        if (currentTime - this.map[existingKey][i] > this.window) {
+          this.map[existingKey].splice(i, 1);
+          --i;
+        }
+      }
+
+      if (this.map[existingKey].length == 0) {
+        delete this.map[existingKey];
+      }
+    }
+
+    if (
+      typeof this.map[key] !== "undefined" &&
+      this.map[key].length >= this.limitPerWindow
+    ) {
+      throw new TooManyRequests();
+    }
+
+    if (typeof this.map[key] === "undefined") {
+      this.map[key] = [];
+    }
+
+    this.map[key].push(currentTime);
+  }
+}
```

Comment:
I'm a _bit_ concerned that each request expands the cardinality of in-memory limit map and that each subsequent request must iterate through all such keys to clean up. This is _probably_ OK because `limit(key)` is only called in cases where `key` is known and _would_ be invoked, and so this shouldn't introduce an actual DOS vector.

But perhaps we can take spilt cleanup routine out of the hot path, and make it asynchronous? The following (totally untested) example might be a way to do this?

```ts
// Max of 10 invocations every 5 seconds (for example)
const max = 10;
const interval = 5000;

// Instead of storing the whole key's history in memory, we just increment a
// counter.
const map: { [key: string]: number } = { __proto__ = null as any };

function limit(k) {
  const v = (map[k] ?? 0) + 1;
  if (v > 10) {
    throw new Error("Rate limit exceeded.").
  }

  // Only write if not limited, to prevent extending the limit unnecessarily.
  map[k] = v;
}

// Handle counter "decay" outside of the normal `limit` calls. If this were a
// multi-threaded language we would do this in a separate thread, but just
// pulling this out of the hot path is probably good enough here.
const i = setInterval(() => {
  for (const k in map) {
    const v  = map[k] - max;
    if (v < 1) {
      delete map[k];
    } else {
      map[k] = v;
    }
  }
}, interval);

// The main complexity is treating this as a long-lived routine, since canceling
// the interval must be part of a graceful shutdown (and ideally _after_ other
// pieces have shut down).
function shutdown() {
  cancelInterval(i)
}
```

If you see any issues with this approach, I'm actually OK going with your original approach at least for now, since in practice I think it's unlikely for these maps to grow so large that all the extra iteration has any noticeable performance impact... it just caught my eye.

Reply:
Hmm, I see your point but I'd prefer to avoid intervals if possible, as they tend to make it harder to test code reliably in my experience. The other issue is that the "resetting" strategy makes it slightly harder to reason about the limits, since if the user is limited to 100 per 60 seconds, they could do 190 in 70 seconds.

If performance in the request is an issue, we could make the following changes:
* When deleting timestamps, we know they're in order, so all we really need to do is find the first one to keep, then delete all timestamps before that.
* In the request, only check if the current caller is over their limit. If they are, delete expired timestamps and check again. This should mean that the vast majority of request calls are a simple length check on a single array.
* On some random percentage of calls, trigger a full "garbage collection" after the call has completed using `setTimeout(..., 0)`. This should prevent the "garbage collection" from slowing down the call itself. While this would make testing harder, we could still set the "garbage collection chance" to 100% in tests, which would get us most of the way there.

It is true that this would use more memory than either approaches, so if that's a major issue we could switch over to the counter/interval approach. I think we could make it work for testing with a little extra setup.

Reply:
Ya, I _completely_ agree with you about the interval being a testing nuisance. That was my biggest hesitation with my suggestion, and I like your suggestion to only GC on a percentage of calls, although as you metioned "randomness" is also a pain. Instead I suppose we could increment a counter and GC every `n` limit calls or on the next call after `n` ms since the last GC.

But you also raise a good point about the bucket sizes, especially in the case of longer intervals. That could probably be alleviated by always using 1-second decay (id GC) intervals and decaying in prorated amounts of `max / (interval / 1000)`. Using counters would reduce memory use compared to the arrays, but I don't think that's a realistic concern here.

---

After thinking about this a bit more, I think I actually like your current approach the best: while _theoretically_ a bit unoptimized for time complexity, it's functionally very simple and clear, and even in its current state it's "protecting" something that is _so_ much more costly that it's only a step in the right direction. So I'm going to revoke my hesitation here: let's go with exactly what you originally wrote, and _if_ it needs to be optimized later (which it almost certainly won't) we have some good possible strategies 👍🏼